src/Disks/IO/CacheableReadBufferFromRemoteFS.cpp

```diff
@@ -0,0 +1,353 @@
+        && state != FileSegment::State::PARTIALLY_DOWNLOADED_NO_CONTINUATION)
+        return;
+
+    auto current_read_range = (*current_file_segment_it)->range();
+    auto last_downloaded_offset = (*current_file_segment_it)->downloadOffset();
+
+    if (file_offset_of_buffer_end > last_downloaded_offset)
+    {
+        impl = reader;
+
+        if ((*current_file_segment_it)->state() == FileSegment::State::PARTIALLY_DOWNLOADED)
+            read_type = ReadType::REMOTE_FS_READ_AND_DOWNLOAD;
+        else
+            read_type = ReadType::REMOTE_FS_READ;
+
+        impl->setReadUntilPosition(current_read_range.right + 1); /// [..., range.right]
+        impl->seek(file_offset_of_buffer_end, SEEK_SET);
+    }
+}
+
+bool CacheableReadBufferFromRemoteFS::nextImpl()
+{
+    if (!initialized)
+        initialize(file_offset_of_buffer_end, getTotalSizeToRead());
+
+    if (current_file_segment_it == file_segments_holder->file_segments.end())
+        return false;
+
+    if (impl)
+    {
+        auto current_read_range = (*current_file_segment_it)->range();
+        auto current_state = (*current_file_segment_it)->state();
+
+        assert(current_read_range.left <= file_offset_of_buffer_end);
+
+        if (file_offset_of_buffer_end > current_read_range.right)
+        {
+            if (!completeFileSegmentAndGetNext())
+                return false;
+        }
+
+        if (current_state == FileSegment::State::PARTIALLY_DOWNLOADED
+                || current_state == FileSegment::State::PARTIALLY_DOWNLOADED_NO_CONTINUATION)
+        {
+            checkForPartialDownload();
+        }
+    }
+    else
+    {
+        impl = createReadBuffer(*current_file_segment_it);
+
+        /// Seek is required only for first file segment in the list of file segments.
+        impl->seek(file_offset_of_buffer_end, SEEK_SET);
+    }
+
+    auto current_read_range = (*current_file_segment_it)->range();
+    size_t remaining_size_to_read = std::min(current_read_range.right, read_until_position - 1) - file_offset_of_buffer_end + 1;
+
+    assert(current_read_range.left <= file_offset_of_buffer_end);
+    assert(current_read_range.right >= file_offset_of_buffer_end);
+
+    if (use_external_buffer)
+    {
+        assert(!internal_buffer.empty());
+        swap(*impl);
+    }
+    else
+    {
+        impl->position() = position();
+        assert(!impl->hasPendingData());
+    }
+
+    bool result;
+    auto & file_segment = *current_file_segment_it;
+    auto download_current_segment = read_type == ReadType::REMOTE_FS_READ_AND_DOWNLOAD;
+
+    try
+    {
+        result = impl->next();
+
+        if (result && download_current_segment)
+        {
+            size_t size = impl->buffer().size();
+
+            if (file_segment->reserve(size))
+                file_segment->write(impl->buffer().begin(), size);
+            else
+                file_segment->complete(FileSegment::State::PARTIALLY_DOWNLOADED_NO_CONTINUATION);
```

Comment:
Add logging about it? Something like "cache is full and we cannot reserve anything, bypassing it".